src/OrchardCore.Modules/OrchardCore.Setup/Filters/AutoSetupStartupFilter.cs

```diff
@@ -0,0 +1,164 @@
+using System;
+using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+using System.Data;
+using System.Data.Common;
+using System.Linq;
+using System.Text;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Localization;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using OrchardCore.Environment.Shell;
+using OrchardCore.Environment.Shell.Scope;
+using OrchardCore.Setup.Options;
+using OrchardCore.Setup.Services;
+
+namespace OrchardCore.Setup
+{
+    public class AutoSetupStartupFilter : IStartupFilter
+    {
+        private readonly IServiceProvider _provider;
+        private readonly IShellHost _shellHost;
+        private readonly ILogger _logger;
+        private readonly IStringLocalizer T;
+
+        public AutoSetupStartupFilter(IServiceProvider provider, IShellHost shellHost, IStringLocalizer<AutoSetupStartupFilter> stringLocalizer, ILogger<AutoSetupStartupFilter> logger)
+        {
+            T = stringLocalizer;
+            _provider = provider;
+            _shellHost = shellHost;
+            _logger = logger;
+        }
+
+        public Action<IApplicationBuilder> Configure(Action<IApplicationBuilder> next)
+        {
+            var shellServices = ShellScope.Context.ServiceProvider;
+            var scopedServices = ShellScope.Services;
+            var settings = ShellScope.Context.Settings;
+            var state = ShellScope.Context.Settings.State;
```

Comment:
Okay looks good

But the scoped services can be also used to resolve shell services, so seems that you would only need

            var services = ShellScope.Services; <= scoped services
            var settings = ShellScope.Context.Settings;

Note: Here we can't inject the scoped services through the ctor, but we could have injected `ShellSettings` that is a tenant singleton, but that's okay because we can get it from the shell scope.

And then below you can check directly

            if (settings.State == TenantState.Uninitialized)
            {

Reply:
Just to say that when we know the context and how to get the right service provider, it is important to use **only one service provider**, here you can just use `services = ShellScope.Services`, no need to use the shell / tenant level one.